### ISSUE TITLE
NixOS: Handle lvm cached volumes properly

### ISSUE THREAD

**Issue by viric** [NixOS/nixpkgs#15516]
There is some work to do to get proper activation of lvm cached volumes in initrd. I currently have a rootfs cached lv, so I care.

When activating the lv (vgchange -a y in our stage1) it wants to call /usr/sbin/cache_check (from thin_provisioning_tools) or it refuses to activate it. I tested this on stage1, but may affect stage2 as well.

Currently I implemented a quick hack in this commit: https://github.com/viric/nixpkgs/commit/497d5e0b1fd5be2af39b91a3f2fe1df7b5d9c85d

But there may be other options:
* Make lvm2 find thin_provisioning_tools. That may help only for stage 2, and it makes lvm2 depend on them.
* Set a lvm.conf file in initrd that sets the absolute path to cache_check

A new nixos module would help, too, that included the modules "dm-cache", "dm-cache-smq", "dm-cache-mq", "dm-cache-cleaner" to stage 1.

What do you think? How to proceed? @aszlig @domenkozar @edolstra

**Comment by dwe11er** [comment]
If we are at it, you could expand it to thin_* and era_* tools (for dm-thin-pool and dm-era modules), both provided by thin_provisioning_tools.

**Comment by viric** [comment]
The #14394 fix broke my setup and I could not boot NixOS 17.03. I used packageOverrides on lvm2 to get it working again.

**Comment by ilya-kolpakov** [comment]
As @dwe11er pointed out the same issue occurs when `vgchange -a y` stumbles upon thin pools in stage 1 and wants to call `thin_check` from the same package.

**Comment by ilya-kolpakov** [comment]
I think a clean solution would be to create an config option that would hold the list of executables and their aliases needed in stage1. The build script would copy them to the image and check that they at least run there. Looking at [filesystem support modules](https://github.com/NixOS/nixpkgs/tree/release-17.09/nixos/modules/tasks/filesystems) it seems to be a common operation that needs to be refactored out.

**Comment by Ekleog** [comment]
Currently having what I believe is the same issue: adding
```
boot.initrd.kernelModules = [ "dm-thin-pool" ];
```
triggers
```
stage-1-init: /nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-thin-provisioning-tools-0.6.3/bin/thin_check: execvp failed: No such file or directory
```

I guess that's due to [this `nuke-refs`](https://github.com/NixOS/nixpkgs/blob/56904d7c423f2b13b37fbd29f39bbb4b52bc7824/pkgs/build-support/kernel/modules-closure.sh#L32).

I wonder if adding `-e ${thin-provisioning-tools}` or equivalent to the `nuke-refs` call would solve the issue, though unfortunately won't have much time to test for the time being.

**Comment by isomarcte** [comment]
I am also having the same issue as @Ekleog. Are there any workarounds for this?

**Comment by Ekleog** [comment]
As I don't have an “essential for boot” filesystem on a thin volume, my workaround up to now (until I or someone else will have time to look into this better) has been adding `options = [ "nofail" ];` to my `fileSystems."/fubar"`, and running `vgchange -ay` on each boot to bring them up. I guess it'd be possible to add the `vgchange -ay` directly in a systemd service, but I avoided doing it in order to make sure I don't forget to look into this sometime :)

**Comment by SnVIZQ** [comment]
I use the trick posted at #nixos IRC channel quite some time ago. The post contained [the link to the configuration.nix snippet](https://paste2.org/MyW3EjYO). It puts the necessary stuff to the initrd image so that lvmcache(7) works:
``` nix
  boot.initrd.extraUtilsCommands = ''
    # Put thin-provisioning-tools into extra-utils and patch lvm accordingly.
    # NOTE: this works only because thin-provisioning-tools string, including
    # version, is longer than extra-utils string. The difference is zeroed. If
    # it would be vice versa there is a chance it would not work because the 
    # stuff after the full path to the tool would be overwritten. Although there
    # seem to be some other, documentation, string just behind the full path
    # name which might not be that important... Anyways, not spending time
    # to figure out how to avoid the patching in case it is not possible doing
    # the proper way.
    for BIN in ${pkgs.thin-provisioning-tools}/bin/*; do
      copy_bin_and_libs $BIN
      SRC="(?<all>/[a-zA-Z0-9/]+/[0-9a-z]{32}-[0-9a-z-.]+(?<exe>/bin/$(basename $BIN)))"
      REP="\"$out\" . \$+{exe} . \"\\x0\" x (length(\$+{all}) - length(\"$out\" . \$+{exe}))"
      PRP="s,$SRC,$REP,ge"
      ${pkgs.perl}/bin/perl -p -i -e "$PRP" $out/bin/lvm
    done
  ''; 
  boot.initrd.extraUtilsCommandsTest = ''
    # The thin-provisioning-tools use pdata_tools binary as a link target of
    # supported utils so it is enough to check only one, the others should
    # "just" work...
    $out/bin/pdata_tools cache_check -V
  '';
  boot.initrd.availableKernelModules = [ "xhci_pci" "ahci" "nvme" "usb_storage" "usbhid" "sd_mod" "dm_persistent_data" "dm_bio_prison" "dm_bufio" "libcrc32c" "crc32c_generic" "dm_cache_smq" ];
  boot.initrd.kernelModules = [ "dm_cache" ];
```

**Comment by tm112358** [comment]
What is the status here?

**Comment by arttuys** [comment]
Anything new so far?

I ran into this same issue, and implemented an ad-hoc patch on my system using YorikSar's workaround [https://github.com/NixOS/nixpkgs/pull/46541#issuecomment-420433116](here). The workaround has worked fine so far, but ideally I'd use an official solution as to avoid accidentally breaking it during an upgrade.

**Comment by Thesola10** [comment]
Not stale, still can't boot my lvm-cache system unattended

**Comment by cizra** [comment]
@Thesola10 I can - here's what I needed to add:

```
# get rid of scary warning about missing cache_check
services.lvm.boot.thin.enable = true;

# if you don't have enough kernel modules, you'll get this error message:
# cache: Error creating cache's policy
boot.initrd.kernelModules = [ "dm-cache" "dm-cache-smq" "dm-cache-mq" "dm-cache-cleaner" ];
boot.kernelModules = [ "kvm-amd" "dm-cache" "dm-cache-smq" "dm-persistent-data" "dm-bio-prison" "dm-clone" "dm-crypt" "dm-writecache" "dm-mirror" "dm-snapshot"];

 # You need to specify preLVM = false, as it defaults to true
 boot.initrd.luks.devices."decrypted".preLVM = false;
 };
```

**Comment by Thesola10** [comment]
@cizra yep, eventually found the relevant modules on a different question about Ubuntu, thanks for the heads-up though.

I still think there should be config options for these module sets (`boot.lvm.enableCache` maybe?) -- after all, isn't it part of the point of NixOS?

**Comment by agowa** [comment]
Getting the same warning about "/usr/sbin/cache_check" being missing but for a non-boot volume that I tried to mount manually once the system was up and running.

running `vgchange -ay` causes:
> /usr/sbin/cache_check: execvp failed: No such file or directory
> WARNING: Check is skipped, please install recommended missing binary /usr/sbin/cache_check!

I suppose one of the LVM2 kernel modules may need to be patched or recompiled for it to look for cache_check somewhere else, as I keep getting this even after `pkgs.thin-provisioning-tools` is installed as a system package (which places it at `/run/current-system/sw/bin/cache_check`)

/cc @7c6f434c @ajs124 Do you know more about this issue by change?

**Comment by ajs124** [comment]
`cache_check` should be available and configured in the initrd, if you set `services.lvm.boot.thin.enable` and `boot.initrd.services.lvm.enable`.
For the regular system, `services.lvm.boot.thin.enable` should be enough.

See here: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/tasks/lvm.nix#L163

**Comment by agowa** [comment]
Thanks, but the link you posted looks like it places the binaries in `/bin/` are you sure this helps to solve this issue? The error message says it is looking for it in `/usr/sbin`. Is the error message just inaccurate and it actually looks for them in multiple places or do we need to patch something else for it to use the correct paths?

(I can't remount the filesystem to test right now, but I'll just try as soon as I can)

**Comment by ajs124** [comment]
The code configures `/etc/lvm/lvm.conf` with absolute paths to these tools, so it'll look something like this, in the resulting system:
```
global/thin_check_executable = /nix/store/c2hjapacwhx6x2765bg5ykspa2iryc4n-thin-provisioning-tools-1.3.0/bin/thin_check
global/thin_dump_executable = /nix/store/c2hjapacwhx6x2765bg5ykspa2iryc4n-thin-provisioning-tools-1.3.0/bin/thin_dump
global/thin_repair_executable = /nix/store/c2hjapacwhx6x2765bg5ykspa2iryc4n-thin-provisioning-tools-1.3.0/bin/thin_repair
global/cache_check_executable = /nix/store/c2hjapacwhx6x2765bg5ykspa2iryc4n-thin-provisioning-tools-1.3.0/bin/cache_check
global/cache_dump_executable = /nix/store/c2hjapacwhx6x2765bg5ykspa2iryc4n-thin-provisioning-tools-1.3.0/bin/cache_dump
global/cache_repair_executable = /nix/store/c2hjapacwhx6x2765bg5ykspa2iryc4n-thin-provisioning-tools-1.3.0/bin/cache_repair
dmeventd/executable = "/nix/store/zn40lda5d8w4gd97gk60mp94yz1p61q7-lvm2-with-dmeventd-2.03.35/bin/dmeventd"
```

I don't think I have a cache volume anywhere anymore right now, but I've been using this for thin lvs for years and it works. 

The reason why we're not just patching the lvm2 package to have these paths correctly is that it would significantly increase the closure size.
